src/topology/continuous_function/t0_sierpinski.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2022 Ivan Sadofschi Costa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ivan Sadofschi Costa
+-/
+import topology.order
+import topology.sets.opens
+import topology.continuous_function.basic
+
+/-!
+# Any T0 space embeds in a product of copies of the Sierpinski space.
+
+We consider `Prop` with the Sierpinski topology. If `X` is a topological space, there is a
+continuous map `product_of_mem_opens` from `X` to `opens X → Prop` which is the product of the maps
+`X → Prop` given by `x ↦ x ∈ u`.
+
+The map `product_of_mem_opens` is always inducing. Whenever `X` is T0, `product_of_mem_opens` is
+also injective and therefore an embedding.
+-/
+
+noncomputable theory
+
+namespace topological_space
+
+lemma eq_induced_by_maps_to_sierpinski (X : Type*) [t : topological_space X] :
+  t = ⨅ (u : opens X), topological_space.induced (λ x, x ∈ u) sierpinski_space :=
+le_antisymm
+(le_infi_iff.2 (λ u, continuous.le_induced $ is_open_iff_continuous_mem.1 u.2))
+(is_open_implies_is_open_iff.mp $ λ u h,
+begin
+  apply is_open_implies_is_open_iff.mpr _ u _,
+  { exact topological_space.induced (λ (x : X), x ∈ u) sierpinski_space },
+  { exact infi_le_of_le ⟨u,h⟩ (le_refl _) },
+  { exact is_open_induced_iff'.mpr ⟨{true}, ⟨is_open_singleton_true, by simp [set.preimage]⟩⟩ },
+end)
```

Comment:
Would you mind writing this lemma in this way? I think it would be easier to maintain, and I for one find it easier to follow.
```suggestion
lemma eq_induced_by_maps_to_sierpinski (X : Type*) [t : topological_space X] :
  t = ⨅ (u : opens X), sierpinski_space.induced (∈ u) :=
begin
  apply le_antisymm,
  { rw [le_infi_iff],
    exact λ u, continuous.le_induced (is_open_iff_continuous_mem.mp u.2) },
  { intros u h,
    rw ← generate_from_Union_is_open,
    apply is_open_generate_from_of_mem,
    simp only [set.mem_Union, set.mem_set_of_eq, is_open_induced_iff'],
    exact ⟨⟨u, h⟩, {true}, is_open_singleton_true, by simp [set.preimage]⟩ },
end
```
You can write `(λ x, x ∈ u) ` instead of `(∈ u)` if you want -- this is just a common way to write the predicate.

This depends on the following missing lemma (it should go in `topology/sets/opens.lean` before or after `mem_coe`):
```lean
@[simp] lemma topological_space.opens.mem_mk {x : α} {U : set α} {h : is_open U} :
  @has_mem.mem _ _ opens.has_mem x ⟨U, h⟩ ↔ x ∈ U := iff.rfl
```